### ISSUE TITLE
The `ephemeralForTest` storage engine is reaching its end of life

### ISSUE THREAD

**Issue by LucVidal360** [nodkz/mongodb-memory-server#742]
## Versions

- NodeJS: 16.19.0
- mongodb-memory-server-*: 8.4.2
- mongodb(the binary version): 5.0.14
- mongodb(the js package): 4.5.0
- system: Linux Kubuntu 22.04

package: mongo-memory-server-global

## What is your question?

The `ephemeralForTest` storage engine is reaching its end of life.

### Context

In the project I'm working on, each test cleans its test collections by issuing `collection.deleteMany({})` on every modified collection during the test. All these `deleteMany` are executed in parallel.

When executing many tests, it happens that the mongod server crashes with a segfault.

After opening [an issue in MongoDB's Jira](https://jira.mongodb.org/browse/SERVER-73550), it appears the "ephemeralForTest" engine has become unstable, starting with MongoDB 5.
They even intend to suppress entirely this engine with MongoDB 7. Cf. [this comment](https://jira.mongodb.org/browse/SERVER-73550?focusedCommentId=5175779&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-5175779):

> In 5.0, we replaced the ephemeralForTest storage engine [...] The new implementation [...] had many bugs, like the one you discovered, and in the 7.0 release, we will remove the ephemeralForTest engine entirely

### Workaround

As suggested, I used the `wiredTiger` engine (thanks to mongo-memory-server's option `instance.storageEngine`) and everything works like a charm.
I made a few benchmarks, and the performances are comparable.

### Question

Knowing this, I have a few interrogations/suggestions:
- should this information be documented?
- would it be possible to have a different default engine depending on the mongodb version used?

**Comment by hasezoey** [comment]
> would it be possible to have a different default engine depending on the mongodb version used?

not currently

> should this information be documented?

probably should be documented yes, though personally i have not run into any issues with that yet

---

i also already read that they want to remove this storage-engine (sadly). also there seems to be a separate in-memory engine for enterprise.

when the change occurs and MMS changes its default version (to that version), then the default will also be changed (because otherwise it literally wouldnt work anymore).

if it really becomes required, the default may be changed earlier (though this will only be in a major MMS version).

for now use the mentioned workaround of manually setting the storage engine

**Comment by LucVidal360** [comment]
Thank you for your answer.

The current workaround is fine by me. I mostly intended to inform you about this.

Feel free to close this issue, or change its status, whatever suits you best.

**Comment by hasezoey** [comment]
> Feel free to close this issue, or change its status, whatever suits you best.

for now i will keep it open until at least a documentation note is made about that it may be unstable and that the default will be changed in the future

**Comment by hasezoey** [comment]
a note has been added in beta documnetation at https://nodkz.github.io/mongodb-memory-server/versions/beta/docs/api/interfaces/mongo-memory-instance-opts#helper-type-storageengine

**Comment by alexbjorlig** [comment]
Wait a moment?

Does this means that if I would like to run MongoDB memory server, with version 7, I have to use wiredTiger?

But if I understand wiredTiger correctly, this means that it will not be in-memory then?

**Comment by hasezoey** [comment]
> But if I understand wiredTiger correctly, this means that it will not be in-memory then?

yes that is correct, because with 7.0.0 `wiredTiger` is the only engine supported (at least for the community server version)

i know that the name `mongodb-memory-server` might not be entirely correct anymore, but i think this package will not rebrand

**Comment by sibelius** [comment]
can we release v9 outside beta? mongo 7 is out

**Comment by hasezoey** [comment]
mms 9 i currently in beta to get some more feedback, but yes, i think it is kinda final (it works in both typegoose and mongoose, my main real-world testing use-cases)

i will try to release it this week

**Comment by sibelius** [comment]
writeTiger is slower than ephemeral and breaks some tests

**Comment by hasezoey** [comment]
well `wiredTiger` is definitely slower than the in-memory engine, but it should not cause different behavior where the in-memory engine worked.

it is recommended to set the db-path to somewhere which is using `tmpfs`(or equivalent)

**Comment by clemp6r** [comment]
> i know that the name mongodb-memory-server might not be entirely correct anymore, but i think this package will not rebrand

The project description and documentation should at least be updated because it is currently misleading.

**Comment by hasezoey** [comment]
> The project description and documentation should at least be updated because it is currently misleading.

Thanks for pointing this out, completely forgot about that, for now i have updated the github description, other places (like documentation landing page) may still need to be updated.

I am not a good writer, so for now the more concise and less miss-leading text is:

> Manage & spin up mongodb server binaries with zero(or slight) configuration for tests.

**Comment by Climax777** [comment]
I know the issue is closed, but wanted to discuss the option of using Percona's Memory Engine? It is a drop-in replacement for MongoDB. 

https://docs.percona.com/percona-server-for-mongodb/7.0/inmemory.html
https://www.percona.com/downloads provides a few links. Not sure if it will follow the same download granularity though.

But I believe it will address the concern relating to performance.

**Comment by sibelius** [comment]
does percona works with mongo 8 ?

is it faster ?

is it an in-place replacement ?

**Comment by sibelius** [comment]
should we open an issue to support percona ?

**Comment by hasezoey** [comment]
@sibelius 
> what about using mongodb in memory engine ?

> An [In-Memory storage engine](https://www.mongodb.com/docs/manual/core/inmemory/#std-label-storage-inmemory) is available in MongoDB Enterprise. Rather than storing documents on-disk, it retains them in-memory for more predictable data latencies.

[source](https://www.mongodb.com/docs/manual/core/storage-engines/)
TL;DR: The `inMemory` engine in official mongodb is not available in the community edition.

> is it an in-place replacement ?

> Percona Server for MongoDB is an enhanced, fully compatible, source available, **drop-in replacement for MongoDB 7.0 Community Edition** with [enterprise-grade features](https://docs.percona.com/percona-server-for-mongodb/7.0/comparison.html).

[source](https://docs.percona.com/percona-server-for-mongodb/7.0/index.html)

> does percona works with mongo 8 ?

From what i have seen on their website, the highest version is currently 7.0.x.

> should we open an issue to support percona ?

I currently do not plan to add this myself, but PRs are welcome.
(Also #701 had been closed as it was a question of "can we use"(note not "support") and i had written that it should work with `SYSTEM_BINARY` and then had become stale)
Suggestion: open a ["Idea" discussion](https://github.com/typegoose/mongodb-memory-server/discussions/categories/ideas) instead of a issue.

**Comment by sibelius** [comment]
I'd tried with download dir and also system binary neither worked

**Comment by hasezoey** [comment]
> I'd tried with download dir

[config option `DOWNLOAD_DIR`](https://typegoose.github.io/mongodb-memory-server/docs/api/config-options#download_dir) changes the directory the binary is downloaded into, it does not set the download URL.

>  also system binary neither worked

what is the [Debug output](https://typegoose.github.io/mongodb-memory-server/docs/guides/enable-debug-mode) if you run it?

**Comment by sibelius** [comment]
```
MongoMS:DryMongoBinary getSystemPath: found system binary path at "percona/percona-server-mongodb-7.0.15-9-x86_64.bookworm-minimal/bin/mongod" +0ms
  MongoMS:DryMongoBinary locateBinary: env "SYSTEM_BINARY" was provided with value: "percona/percona-server-mongodb-7.0.15-9-x86_64.bookworm-minimal/bin/mongod" +0ms
  MongoMS:DryMongoBinary getSystempath +0ms
  MongoMS:DryMongoBinary getSystemPath: found system binary path at "percona/percona-server-mongodb-7.0.15-9-x86_64.bookworm-minimal/bin/mongod" +0ms
  MongoMS:MongoBinary getPath: Spawning binaryPath "percona/percona-server-mongodb-7.0.15-9-x86_64.bookworm-minimal/bin/mongod" to get version +1ms
  MongoMS:MongoMemoryServer Mongo[unknown]: _startUpInstance threw a Error:  TypeError: Cannot read properties of undefined (reading 'toString')
    at MongoBinary.getPath (woovi/node_modules/mongodb-memory-server-core/lib/util/MongoBinary.js:79:28)
    at async MongoInstance.start (woovi/node_modules/mongodb-memory-server-core/lib/util/MongoInstance.js:135:26)
    at async MongoInstance.create (woovi/node_modules/mongodb-memory-server-core/lib/util/MongoInstance.js:84:9)
    at async MongoMemoryServer._startUpInstance (woovi/node_modules/mongodb-memory-server-core/lib/MongoMemoryServer.js:223:26)
    at async MongoMemoryServer.start (woovi/node_modules/mongodb-memory-server-core/lib/MongoMemoryServer.js:84:9)
    at async MongoMemoryServer.create (woovi/node_modules/mongodb-memory-server-core/lib/MongoMemoryServer.js:63:9)
    at async getMMSMongoD (woovi/packages/testutils/src/config/getMMSMongoD.js:16:20)
    at async module.exports (woovi/packages/testutils/src/config/mongo/setup.js:20:20)
    at async woovi/node_modules/@jest/core/build/runGlobalHook.js:109:13
    at async waitForPromiseWithCleanup (woovi/node_modules/@jest/transform/build/ScriptTransformer.js:160:5) +3ms
  MongoMS:MongoMemoryServer Mongo[unknown]: stop: Called .stop() method +0ms
  MongoMS:MongoMemoryServer Mongo[unknown]: stop: "instanceInfo" is not defined (never ran?) +0ms
TypeError: Cannot read properties of undefined (reading 'toString')
    at MongoBinary.getPath (woovi/node_modules/mongodb-memory-server-core/lib/util/MongoBinary.js:79:28)
    at async MongoInstance.start (woovi/node_modules/mongodb-memory-server-core/lib/util/MongoInstance.js:135:26)
    at async MongoInstance.create (woovi/node_modules/mongodb-memory-server-core/lib/util/MongoInstance.js:84:9)
    at async MongoMemoryServer._startUpInstance (woovi/node_modules/mongodb-memory-server-core/lib/MongoMemoryServer.js:223:26)
    at async MongoMemoryServer.start (woovi/node_modules/mongodb-memory-server-core/lib/MongoMemoryServer.js:84:9)
    at async MongoMemoryServer.create (woovi/node_modules/mongodb-memory-server-core/lib/MongoMemoryServer.js:63:9)
    at async getMMSMongoD (woovi/packages/testutils/src/config/getMMSMongoD.js:16:20)
    at async module.exports (woovi/packages/testutils/src/config/mongo/setup.js:20:20)
    at async woovi/node_modules/@jest/core/build/runGlobalHook.js:109:13
    at async waitForPromiseWithCleanup (woovi/node_modules/@jest/transform/build/ScriptTransformer.js:160:5)
```

**Comment by hasezoey** [comment]
Thanks for the output, i think i should do more checking there, though from what i know this should never happen in nodejs (according to types):
https://github.com/typegoose/mongodb-memory-server/blob/a9d8986064ab7fee4a33a7f38375a24aa20de69c/packages/mongodb-memory-server-core/src/util/MongoBinary.ts#L91-L92

Does anything change if you provide a absolute path to the binary? (ie `/path/to/somewhere/mongod`)

**Comment by sibelius** [comment]
this is the absolute path, I only removed in the logs

**Comment by hasezoey** [comment]
> this is the absolute path, I only removed in the logs

just to confirm the path you had set for `SYSTEM_BINARY` was starting with a `/`(absolute)? (ie `/percona/percona-server-mongodb-7.0.15-9-x86_64.bookworm-minimal/bin/mongod` instead of `percona/percona-server-mongodb-7.0.15-9-x86_64.bookworm-minimal/bin/mongod`)

**Comment by sibelius** [comment]
it was absolute

```
/Users/sibeliusseraphini/Dev/entria/woovi-projects/percona/percona-server-mongodb-7.0.15-9-x86_64.bookworm-minimal/bin/mongod
```

**Comment by hasezoey** [comment]
Thanks, then it is very weird that nodejs seemingly did not throw a error and gave back what it shouldnt have.
In any case, i have now implemented a path absolutization for the SYSTEM_BINARY (in beta) and also a extra check that `stdout` is defined.

Could you manually try running the percona binary with argument `--version`? (if it works, please provide the output)
Additionally, you can try [`10.1.3-beta.1`](https://github.com/typegoose/mongodb-memory-server/releases/tag/v10.1.3-beta.1) with [config option `SYSTEM_BINARY_VERSION_CHECK`](https://typegoose.github.io/mongodb-memory-server/docs/api/config-options#system_binary_version_check) set to `false` to see if skipping the version check makes it work.

**Comment by sibelius** [comment]
You can download percona and give it a try

I will test more

**Comment by sibelius** [comment]
try this https://github.com/sibelius/mongodb-memory-server-percona
